### ISSUE TITLE
[SuperEditor] - Limit undo stack history with Editable snapshots

### ISSUE THREAD

**Issue by matthew-carroll** [superlistapp/super_editor#2154]
Currently the undo stack grows without bounds. It could be dozens, hundreds, or thousands of changes long.

This unbounded length introduces two possible issues:
 * Excessive memory consumption.
 * Degrading performance when running undo's.

We should come up with a strategy, or a collection of strategies, for collapsing the undo stack down into non-undo-able Editable states.

First, this requires adding the ability of each Editable to create snapshot representations of itself, and restore itself from the snapshot.

Second, we need to keep in mind that once the undo events are collapsed, they're unrecoverable.

**Comment by KevinBrendel** [comment]
In theory, there could be some more complex collapsing strategies, where e.g. depending on how far into the past the stack goes, it could become more sparse. Either collapsing it down to every second, third etc. state, or trying to make it time based where there would be a state preserved per time interval (e.g. every minute, every ten minutes, etc.).

But for our use-case a very simple strategy would suffice, where x number of undo steps are possible at any one time and anything older than that is discarded immediately. If the stack is full, this would result in the oldest state being discarded each time a new step is added. If this is significantly processing intensive, maybe an optional strategy could allow the app to manually trigger the collapsing in an appropriate moment (rather than slowing down each editing step).

Depending on the use-case, it might make sense to try and limit the stack not on a fixed amount of steps, but based on the amount of memory it consumes. For example, if each step requires keeping a 100mb image in memory, we could have fewer steps than if each step is just adding some text characters. But this probably isn't necessary for our use-case, as it is more likely to only have a few images that are manipulated multiple times.. so each state should only have a reference to the image, not store a copy of it.

Not sure how exactly the undo trees play into all of this. We will only use one branch for now. Probably might have to be able to configure all of this either for each branch, or globally across all branches.

In general, being able to define callbacks (or implementing an interface/creating a subclass) which receive all of the required information as parameters and return the desired result/can perform the required actions via public functions might be the easiest way to accommodate the different use-cases. And then use the same pattern as elsewhere where a few default strategies are already implemented.

**Comment by matthew-carroll** [comment]
Thanks @KevinBrendel - I think I'll try to create a pluggable object and then create an initial implementation of an X-undo length. I'm also thinking that rather than collapsing on every change, we do something like when we get 100 history items, we collapse the oldest 50 and then expand again. That way we're not constantly collapsing.

BTW, have we discussed your use-case before? I wasn't sure. If not, can you tell us a bit about how you're making use of Super Editor?

**Comment by KevinBrendel** [comment]
Good idea to give it some leeway so it doesn't have to collapse as often!

We are using Super Editor to build a new version of QuikFlow. Each node will contain a Super Editor, thereby enabling support for rich text, links, tasks, images, and other content types in the future. For the current version we have already switched to SuperTextField to make sure that e.g. Korean is still properly supported (there are some old issues stating that Super Editor does not support it, but it seems to work in the current versions from our testing.)

<img width="1335" alt="Screenshot 2024-07-10 at 22 53 56" src="https://github.com/superlistapp/super_editor/assets/55455206/8e6a7b22-e455-4469-bdd8-ee3011adf862">
